esp/services/ws_rrd/RRDGraphWrapper.cpp

```diff
@@ -0,0 +1,287 @@
+
+    bool bClusterSpecified = true;
+    if (nClusterCount == 0)
+    {
+         bClusterSpecified = false;
+    }
+    bool bServerSpecified = true;
+    if (nServerCount == 0)
+    {
+         bServerSpecified = false;
+    }
+
+    if (bServerSpecified == true && bClusterSpecified == false) // invalid
+    {
+        PROGLOG("ganglia-monitoring: Cannot specifiy server with no cluster");
+        throw MakeStringException(0, "ganglia-monitoring: Cannot specifiy server with no cluster");
+    }
+
+    for (int idxMetric = 0; idxMetric < nRRDFileCount; idxMetric++)
+    {
+        for (int idx = 0; (idx < nClusterCount || bClusterSpecified == false) && (idx < nServerCount || bServerSpecified == false) && idxMetric < nRRDFileCount && idx < nPlotColors; idx++)
+        {
+            const char *pServer = bServerSpecified ? strArrServers.item(idx) : "";
+            const char *pCluster = bClusterSpecified ? strArrClusters.item(idx) : "";
+            const char *pRRDFile = strArrRRDFiles.item(idxMetric);
+
+            if ((pServer == NULL || strcmp(pServer, pSummaryInfo) == 0 || pServer[0] == 0) && (pCluster == NULL || *pCluster == 0))
+            {
+                strRRDFilePath.set(pDefaultPathSummaryInfo);
+            }
+            else
+            {
+                strRRDFilePath.set(pDefaultPathRRDRoot);
+                strRRDFilePath.appendf("%s/", pCluster);
+
+                if (pServer != NULL && *pServer != 0)
+                {
+                    strRRDFilePath.append(pServer);
+                }
+                else
+                {
+                    strRRDFilePath.append(pSummaryInfo);
+                }
+                strRRDFilePath.append("/");
+            }
+            strRRDFilePath.appendf("%s.rrd", pRRDFile);
+
+            if (checkFileExists(strRRDFilePath.str()) == false)
+            {
+                PROGLOG("Can not find rrd file for '%s'", strRRDFilePath.str());
+                throw MakeStringException(0, "Can not find rrd file for '%s'", strRRDFilePath.str());
+            }
+
+            strRRDGraphCmd.appendf(" %s%d%d=%s", pDataDefinitionPrefix1, idxMetric+1, idx+1, strRRDFilePath.str());
+            strRRDGraphCmd.appendf("%s", pDataDefinitionSuffix);
+            strRRDGraphCmd.appendf(" %s%d:ds%d%d#%s:%s", pLineDefinition, nLineCount++, idxMetric+1, idx+1, PlotColor[(nLineCount-1)%nPlotColors], pRRDFile);
+
+            if (bServerSpecified == false || bClusterSpecified == false)
+            {
+                break;
+            }
+        }
+    }
+
+    strRRDGraphCmd.appendf(" %s %d", pWidth, nWidth);
+    strRRDGraphCmd.appendf(" %s %d", pHeight, nHeight);
+
+    if (pTitle != NULL && *pTitle != 0)
+    {
+        strRRDGraphCmd.appendf(" -t '%s'", pTitle);
+    }
+
+    PROGLOG("RRDTOOL GRAPH CMD --> %s <--", strRRDGraphCmd.str());
+    FILE *fp = popen(strRRDGraphCmd.str(), "r");
+
+    if (fp == NULL)
+    {
+        return NULL;
+    }
+
+    pclose(fp);
```

Comment:
Before closing the stream, through pclose,  I want to ensure that I have a valid FILE pointer.